File: dotNet-sdk/Semantria/Mapping/Configuration/Configuration.cs
using System;
using System.Runtime.Serialization;
using System.Collections.Generic;
using System.Xml.Serialization;

namespace Semantria.Com.Mapping.Configuration
{
    [DataContract(Name = "configuration")]
    [XmlRootAttribute("configuration", Namespace = "")]
    public class Configuration
    {
        [DataMember(Name = "config_id", EmitDefaultValue=false)]
        [XmlElementAttribute("config_id")]
        public string ConfigId { get; set; }

        [DataMember(Name = "template", EmitDefaultValue = false)]
        [XmlElementAttribute("template")]
        public string Template { get; set; }

        [DataMember(Name = "name")]
        [XmlElementAttribute("name")]
        public string Name { get; set; }

        [DataMember(Name = "one_sentence")]
        [XmlElementAttribute("one_sentence")]
        public bool OneSentence { get; set; }
        
        [DataMember(Name = "is_primary")]
        [XmlElementAttribute("is_primary")]
        public bool IsPrimary { get; set; }

        [DataMember(Name = "auto_responding")]
        [XmlElementAttribute("auto_responding")]
        public bool AutoResponding { get; set; }

        [DataMember(Name = "language")]
        [XmlElementAttribute("language")]
        public string Language { get; set; }

        [DataMember(Name = "chars_threshold", EmitDefaultValue = false)]
        [XmlElementAttribute("chars_threshold")]
        public int CharsThreshold { get; set; }

        [DataMember(Name = "callback")]
        [XmlElementAttribute("callback")]
        public string Callback { get; set; }

        [DataMember(Name = "document")]
        [XmlElementAttribute("document", typeof(DocConfiguration))]
        public DocConfiguration Document { get; set; }

        [DataMember(Name = "collection")]
        [XmlElementAttribute("collection", typeof(CollConfiguration))]
        public CollConfiguration Collection { get; set; }
    }

    [DataContract(Name = "document")]
    [XmlRootAttribute("document", Namespace = "")]
    public class DocConfiguration
    {
        [DataMember(Name = "concept_topics_limit")]
        [XmlElementAttribute("concept_topics_limit")]
        public int ConceptTopicsLimit { get; set; }

        [DataMember(Name = "query_topics_limit")]
        [XmlElementAttribute("query_topics_limit")]
        public int QueryTopicsLimit { get; set; }

        [DataMember(Name = "named_entities_limit")]
        [XmlElementAttribute("named_entities_limit")]
        public int NamedEntitiesLimit { get; set; }

        [DataMember(Name = "user_entities_limit")]
        [XmlElementAttribute("user_entities_limit")]
        public int UserEntitiesLimit { get; set; }

        [DataMember(Name = "entity_themes_limit")]
        [XmlElementAttribute("entity_themes_limit")]
        public int EntityThemesLimit { get; set; }

        [DataMember(Name = "themes_limit")]
        [XmlElementAttribute("themes_limit")]
        public int ThemesLimit { get; set; }

        [DataMember(Name = "phrases_limit")]
        [XmlElementAttribute("phrases_limit")]
        public int PhrasesLimit { get; set; }

        [DataMember(Name = "summary_limit")]
        [XmlElementAttribute("summary_limit")]
        public int SummaryLimit { get; set; }
    }

    [DataContract(Name = "collection")]
    [XmlRootAttribute("collection", Namespace = "")]
    public class CollConfiguration
    {
        [DataMember(Name = "facets_limit")]
        [XmlElementAttribute("facets_limit")]
        public int FacetsLimit { get; set; }

        [DataMember(Name = "facet_atts_limit")]
        [XmlElementAttribute("facet_atts_limit")]
        public int FacetAttsLimit { get; set; }

        [DataMember(Name = "concept_topics_limit")]
        [XmlElementAttribute("concept_topics_limit")]
        public int ConceptTopicsLimit { get; set; }

        [DataMember(Name = "query_topics_limit")]
        [XmlElementAttribute("query_topics_limit")]
        public int QueryTopicsLimit { get; set; }

        [DataMember(Name = "named_entities_limit")]
        [XmlElementAttribute("named_entities_limit")]
        public int NamedEntitiesLimit { get; set; }

        [DataMember(Name = "themes_limit")]
        [XmlElementAttribute("themes_limit")]
        public int ThemesLimit { get; set; }
    }
}

namespace Semantria.Com.Mapping.Configuration.Stub
{
    [DataContract(Name = "configurations")]
    [XmlRootAttribute("configurations", Namespace = "")]
    public class Configurations
    {
        [DataMember(Name = "configuration")]
        [XmlElementAttribute("configuration")]
        public List<Configuration> Data { get; set; }
    }

    [DataContract(Name = "configurations")]
    [XmlRootAttribute("configurations", Namespace = "")]
    public class ConfigurationManagable : IManagable<Configuration, string>
    {
        private List<Configuration> _added = new List<Configuration>();
        private List<string> _removed = new List<string>();

        #region IManagable<T,U> Members

        [DataMember(Name = "added")]
        [XmlArrayAttribute("added")]
        [XmlArrayItemAttribute("configuration", typeof(Configuration))]
        public List<Configuration> Added
        {
            get
            {
                return _added;
            }
            set
            {
                _added = value;
            }
        }

        [DataMember(Name = "removed")]
        [XmlArrayAttribute("removed")]
        [XmlArrayItemAttribute("configuration", typeof(string))]
        public List<string> Removed
        {
            get
            {
                return _removed;
            }
            set
            {
                _removed = value;
            }
        }

        #endregion
    }
}

File: cpp-sdk/src/objects/configuration/Configuration.cpp
#include "Configuration.h"

Configuration::Configuration() {
    is_primary = false;

    //document
    entity_themes_limit = 0;
    summary_limit = 0;
    phrases_limit = 0;
    doc_themes_limit = 0;
    doc_query_topics_limit = 0;
    doc_concept_topics_limit = 0;
    doc_named_entities_limit = 0;
    user_entities_limit = 0;

    //collection
    facets_limit = 0;
    facet_atts_limit = 0;
    coll_themes_limit = 0;
    coll_query_topics_limit = 0;
    coll_concept_topics_limit = 0;
    coll_named_entities_limit = 0;
}

Configuration::~Configuration() {}

void Configuration::Serialize(Json::Value& root) {
    if (config_id != "") {
        root["config_id"] = config_id;
    }

    if (name != "") {
        root["name"] = name;
    }

    if (templateId != "") {
        root["template"] = templateId;
    }

    root["one_sentence"] = one_sentence;
    root["auto_responding"] = auto_responding;
    root["is_primary"] = is_primary;
    root["language"] = language;
    root["chars_threshold"] = chars_threshold;
    if (callback != "") {
        root["callback"] = callback;
    }

    //document
    root["document"]["entity_themes_limit"] = entity_themes_limit;
    root["document"]["summary_limit"] = summary_limit;
    root["document"]["phrases_limit"] = phrases_limit;
    root["document"]["themes_limit"] = doc_themes_limit;
    root["document"]["query_topics_limit"] = doc_query_topics_limit;
    root["document"]["concept_topics_limit"] = doc_concept_topics_limit;
    root["document"]["named_entities_limit"] = doc_named_entities_limit;
    root["document"]["user_entities_limit"] = user_entities_limit;

    //collection
    root["collection"]["facets_limit"] = facets_limit;
    root["collection"]["facet_atts_limit"] = facet_atts_limit;
    root["collection"]["themes_limit"] = coll_themes_limit;
    root["collection"]["query_topics_limit"] = coll_query_topics_limit;
    root["collection"]["concept_topics_limit"] = coll_concept_topics_limit;
    root["collection"]["named_entities_limit"] = coll_named_entities_limit;
}

void Configuration::Deserialize(Json::Value& root) {
    config_id = root.get("config_id", "").asString();
    name = root.get("name", "").asString();
    one_sentence = root.get("one_sentence", false).asBool();
    auto_responding = root.get("auto_responding", false).asBool();
    is_primary = root.get("is_primary", false).asBool();
    language = root.get("language", false).asString();
    chars_threshold = root.get("chars_threshold", 0).asUInt();
    callback = root.get("callback", "").asString();

    //document
    entity_themes_limit = root.get("entity_themes_limit", 0).asUInt();
    summary_limit = root.get("summary_limit", 0).asUInt();
    phrases_limit = root.get("phrases_limit", 0).asUInt();
    doc_themes_limit = root.get("themes_limit", 0).asUInt();
    doc_query_topics_limit = root.get("query_topics_limit", 0).asUInt();
    doc_concept_topics_limit = root.get("concept_topics_limit", 0).asUInt();
    doc_named_entities_limit = root.get("named_entities_limit", 0).asUInt();
    user_entities_limit = root.get("user_entities_limit", 0).asUInt();

    //collection
    facets_limit = root.get("facets_limit", 0).asUInt();
    facet_atts_limit = root.get("facet_atts_limit", 0).asUInt();
    coll_themes_limit = root.get("themes_limit", 0).asUInt();
    coll_query_topics_limit = root.get("query_topics_limit", 0).asUInt();
    coll_concept_topics_limit = root.get("concept_topics_limit", 0).asUInt();
    coll_named_entities_limit = root.get("named_entities_limit", 0).asUInt();
}
